src/mainboard/sunxi/nezha/src/main.rs

```diff
@@ -37,49 +101,27 @@ pub extern "C" fn _start() -> ! {
     let w = &mut print::WriteTo::new(console);
     writeln!(w, "## Loading payload\r").unwrap();
 
-    // see ../fixed-dtfs.dts
-    // TODO: adjust when DRAM driver is implemented / booting from SPI
-    let mem = 0x4000_0000;
-    let cached_mem = 0x8000_0000;
-    let payload_offset = 0x2_0000;
-    let payload_size = 0x1e_0000;
-    let linuxboot_offset = 0x20_0000;
-    let linuxboot_size = 0x120_0000;
-    let dtb_offset = 0x140_0000;
-    let dtb_size = 0xe000;
-
-    // TODO; This payload structure should be loaded from boot medium rather
-    // than hardcoded.
-    let segs = &[
-        payload::Segment {
-            typ: payload::stype::PAYLOAD_SEGMENT_ENTRY,
-            base: cached_mem,
-            data: &mut SectionReader::new(&Memory {}, mem + payload_offset, payload_size),
-        },
-        payload::Segment {
-            typ: payload::stype::PAYLOAD_SEGMENT_ENTRY,
-            base: cached_mem,
-            data: &mut SectionReader::new(&Memory {}, mem + linuxboot_offset, linuxboot_size),
-        },
-        payload::Segment {
-            typ: payload::stype::PAYLOAD_SEGMENT_ENTRY,
-            base: cached_mem,
-            data: &mut SectionReader::new(&Memory {}, mem + dtb_offset, dtb_size),
-        },
-    ];
     // TODO: Get this from configuration
     let use_sbi = true;
     if use_sbi {
+        decompress(w);
         writeln!(
             w,
             "Handing over to SBI, will continue at 0x{:x}\r",
-            mem + linuxboot_offset
+            LINUXBOOT_ADDR
         )
         .unwrap();
-        sbi_init(mem + linuxboot_offset, mem + dtb_offset);
+        sbi_init(LINUXBOOT_ADDR, DTB_ADDR);
     } else {
-        let entry = mem + payload_offset;
-        let payload: payload::Payload = payload::Payload {
+        // TODO; This payload structure should be loaded from boot medium rather
+        // than hardcoded.
+        let entry = MEM + PAYLOAD_OFFSET;
+        let segs = &[payload::Segment {
+            typ: payload::stype::PAYLOAD_SEGMENT_ENTRY,
+            base: CACHED_MEM,
+            data: &mut SectionReader::new(&Memory {}, entry, PAYLOAD_SIZE),
```

Comment:
See this line? In principle, if the decompressor were a Device, the data here could be a &mut DecompressReader::new(device: memory section reader, or disk, or serial line, depending on jumpers);

There's some power there it would be nice to use. But it's more work for sure ... 

anyway ... I think this should go in as is, because it is the difference between usable and unusable kernel payload?